### PR TITLE
Add ledger object models

### PR DESCRIPTION
**Pull request by LimpidCrypto**
This is the follow up to #57 

### Ledger object models
Adds [ledger object](https://xrpl.org/ledger-object-types.html) models including the `AMM` object model.

### Models in general
It is also an attempt to implement a general order for fields (models/mod.rs):
```md
//! Order of models:
//! 1. Type of model
//! 2. Required common fields in alphabetical order
//! 3. Required specific fields in alphabetical order
//! 4. Optional common fields in alphabetical order
//! 5. Optional specific fields in alphabetical order
```
If agreed on the order:
- [x] open a new issue for refactoring the other models

### `serde_with_tag`
Moved the `Helper` to `fn serialize` and `fn deserialize` to avoid defining it multiple times in a file.

### Context of Change

#27 

### Type of Change

- [x] New feature (non-breaking change which adds functionality)

## Test Plan

- [x] serialization tests
- [ ] deserialization tests

## Future Tasks

Keep an eye on the [`XRPFees`](https://xrpscan.com/amendment/93E516234E35E08CA689FA33A6D38E103881F8DCB53023F728C307AA89D515A7) amendment. It could change how some fields are typed and/or named.

### CODE REVIEW

File: src/_serde/mod.rs [hunk 1]
```diff
@@ -1,15 +1,85 @@
...
+                    Err(_) => {
+                        return Err(de::Error::custom("SerdeIntermediateStepError: Failed to turn flag into `u32` during deserialization"));
+                    }
+                };
+            }
+            Err(_) => {
+                return Err(de::Error::custom("SerdeIntermediateStepError: Failed to turn flag into `String` during deserialization"));
+            }
+        };
+    }
```
Comment (sephynox) [inline src/_serde/mod.rs:69]: I think an `if let` or pattern binding might make this a bit cleaner.

File: src/_serde/mod.rs [hunk 2]
```diff
@@ -1,15 +1,85 @@
...
+                Err(_) => {
+                    // TODO: Find a way to use custom errors
+                    Err(ser::Error::custom("SerdeIntermediateStepError: Failed to turn flags into `Vec<u32>` during serialization"))
+                }
+            }
+        }
+        Err(_) => Err(ser::Error::custom(
+            "SerdeIntermediateStepError: Failed to turn flags into `Value` during serialization",
+        )),
+    }
```
Comment (sephynox) [inline src/_serde/mod.rs:35]: I think an if let or pattern binding might make this a bit cleaner.

File: src/_serde/mod.rs [hunk 3]
```diff
@@ -32,26 +99,47 @@ pub mod txn_flags {
...
+        match flags_vec_result {
+            Ok(flags_vec) => {
+                if flags_vec.is_empty() {
+                    Ok(None)
+                } else {
+                    Ok(Some(flags_vec))
+                }
             }
+            Err(error) => Err(error),
         }
```
Comment (sephynox) [inline src/_serde/mod.rs:112]: An if/else would be simpler here.

File: src/_serde/mod.rs (hunk 1 above)
Comment (LimpidCrypto) [reply src/_serde/mod.rs:69]: Honestly I find using `match` to resolve a `Result` more idiomatic and consistent.

Comment (sephynox) [reply src/_serde/mod.rs:69]: Ok

**Pull request merged**